### PR TITLE
Interactions

### PR DESCRIPTION
**Pull request by L0neGamer**
Add interactions handling, including making application commands (currently basic but can be expanded fairly easily) and working with interactions.

also updates discord-haskell to the latest version (1.12.4) and fixes the issues there.

Also also adds threads support, and made it so that tablebot automatically joins any threads that are created.

Example working here:

slashcommands

### CODE REVIEW

**Comment by finnbar** [comment]
Very quick question - I know that Sahasrara is planning to eventually be based off this and is a public bot so is present in many servers. How do interactions (ha) interact with this? I haven't read the code through yet, but given Amelie's comment I assume there's no way to turn them off? Would this PR mean that public bots are no longer allowed?

(Admittedly, I think the auth stuff is also partially incompatible with public bots since it relies on fixed role IDs. This is fine if you're happy with authorised commands only being able to be run in a given server though, or if I remember correctly Server Owner gets to bypass the auth stuff anyway. If I'm wrong, that may be another required fix for making the bot easier for others to run.)

**Comment by L0neGamer** [comment]
Currently application commands (slash commands, user commands, message commands, and modals) are only set up in a single, given server (which is required). There are two ways to solve this:

1) the server id for such application command creation is instead a list of ids, specifying which servers to add the commands to. This has the downside that each such server would need to be added to the list, which is infeasible for a public bot. ofc something can be aded which allows a server to opt in, meaning it uses the database to store opt-in servers, but that's a bit excessive because...
2) you can create global application commands.; the bot doesn't use those currently because they take up to an hour to activate, but we can combine the best of both worlds by having an absent server id meaning "create these commands globally", and then having a list present meaning "create in these servers"

Basically, there's a variety of options all of which should be fairly easy to do. 

Another change I should make is to let having these application commands be optional.

File: src/Tablebot/Plugins/Roll/Plugin.hs [hunk 1]
```diff
@@ -181,7 +231,7 @@ statsCommand = Command "stats" statsCommandParser []
               liftDiscord $
                 void $
                   restCall
-                    ( CreateMessageDetailed (messageChannel m) (MessageDetailedOpts (msg range') False Nothing (Just (T.unwords (snd <$> range') <> ".png", toStrict image)) Nothing Nothing)
+                    ( CreateMessageDetailed (messageChannelId m) (MessageDetailedOpts (msg range') False Nothing (Just (T.unwords (snd <$> range') <> ".png", toStrict image)) Nothing Nothing Nothing Nothing)
```
Comment (finnbar) [inline src/Tablebot/Plugins/Roll/Plugin.hs:234]: Could there be a helper function that avoids all of the `Nothing` here, or is there no way to narrow it down? (Fine if not, just seems like a whole lot of Nothing.)

File: src/Tablebot/Plugins/Roll/Plugin.hs [hunk 2]
```diff
@@ -57,22 +62,52 @@ rollDice' e' t m = do
...
+    ( (messageDetailsBasic msg)
+        { messageDetailsComponents =
+            Just
+              [ ComponentActionRowButton
+                  [ (mkButton buttonName (T.take 100 buttonCustomId))
+                      { componentButtonEmoji = Just (Emoji (Just 0) "🎲" Nothing Nothing Nothing (Just False)),
+                        componentButtonDisabled = buttonDisabled
+                      }
+                  ]
+              ]
```
Comment (finnbar) [inline src/Tablebot/Plugins/Roll/Plugin.hs:78]: Could maybe be a helper function `makeInteractionButton`, to avoid much use of Nothing? (Again, it might be that there are serious cases where those aren't Nothing and thus my point is not sensible.)

Also, what's with the `T.take`?

File: src/Tablebot/Plugins/Quote.hs [hunk 3]
```diff
@@ -270,29 +309,158 @@ deleteQ qId m =
...
+      ApplicationCommandOptionSubcommand
+        "edit"
+        "edit a quote"
+        [ ApplicationCommandOptionValueInteger "quoteid" "the id of the quote to edit" True (Left False) Nothing Nothing,
+          ApplicationCommandOptionValueString "quote" "what the actual quote is" False (Left False),
+          ApplicationCommandOptionValueString "author" "who authored this quote" False (Left False)
+        ]
+
+quoteApplicationCommandRecv :: Interaction -> DatabaseDiscord ()
+quoteApplicationCommandRecv i@InteractionApplicationCommand {interactionDataApplicationCommand = InteractionDataApplicationCommandChatInput {interactionDataApplicationCommandOptions = Just (InteractionDataApplicationCommandOptionsSubcommands [InteractionDataApplicationCommandOptionSubcommandOrGroupSubcommand subc])}} = case subcname of
```
Comment (finnbar) [inline src/Tablebot/Plugins/Quote.hs:382]: why is this line so long

File: src/Tablebot/Plugins/Quote.hs [hunk 4]
```diff
@@ -270,29 +309,158 @@ deleteQ qId m =
...
+    editQuoteAppComm =
+      ApplicationCommandOptionSubcommand
+        "edit"
+        "edit a quote"
+        [ ApplicationCommandOptionValueInteger "quoteid" "the id of the quote to edit" True (Left False) Nothing Nothing,
+          ApplicationCommandOptionValueString "quote" "what the actual quote is" False (Left False),
+          ApplicationCommandOptionValueString "author" "who authored this quote" False (Left False)
+        ]
+
+quoteApplicationCommandRecv :: Interaction -> DatabaseDiscord ()
```
Comment (finnbar) [inline src/Tablebot/Plugins/Quote.hs:381]: This definition makes it look like you have to put in a hell of a lot of work to get interactions to work. Is there a way this could be reused/reduced? As I say in the conclusion of this review, I'd like to see a tutorial about interactions and slash commands - maybe when you write this tutorial it'll become clear how things could be simplified.

File: src/Tablebot/Handler.hs [hunk 5]
```diff
@@ -99,3 +113,33 @@ runCron pool (CCronJob delay fn) = do
...
+-- tablebot cache, create the given application commands, purge ones that
+-- weren't created by us, and place the application command id's and their
+-- actions in the cache.
+submitApplicationCommands :: [CompiledApplicationCommand] -> MVar TablebotCache -> DiscordHandler ()
+submitApplicationCommands compiledAppComms cacheMVar =
+  ( do
+      -- generate the application commands, cleaning up any application commands we don't like
+      serverIdStr' <- liftIO $ lookupEnv "SERVER_ID"
+      case serverIdStr' of
+        Nothing -> pure ()
```
Comment (finnbar) [inline src/Tablebot/Handler.hs:127]: Is this correct? I thought a blank string lead to global application commands or something?

**Review by finnbar — CHANGES_REQUESTED** [review]
The code works perfectly when testing, and provides a great user experience. I'm also astounded it still works even after some blacklisting. I have no worries about how it functions.

While I trust your code to be good, I had issues understanding it, however. When you provide a PR with _33_ files changed, you need to explain very clearly what you've implemented. I had to guess what functionality you'd actually implemented because you don't say all of the commands which benefit from slash commands / interactions, only that you can now reroll with an interaction. You also seem to have made a bunch of dice changes with `Proxy`?

As another future PRs thing - it would have been really helpful for you to explain the main types / functions being added with each feature. For example, when you talk about interactions it would have been helpful for you to point out the added types `EnvApplicationCommandRecv` and `EnvComponentRecv` - you don't need to describe them in great detail, just point them out so I know where to go first.

I want a tutorial written about Interactions before I feel comfortable merging this - a number of questions arose while I was reading the code that I think would be aided by a tutorial stepping through how to add interactions and slash commands. This is particularly pertinent because it seems like a plugin writer will have to put in significant effort if they'd like to add application commands to their plugin.

**Comment by finnbar** [comment]
Notes for when I have to get this running in the actual bot:

* Need to set `SERVER_ID` and `stack clean` in the instance running on hopper, since this requires `SERVER_ID` and does a bunch of package changes.
* Need to update Discord Intents to allow message reading.

File: src/Tablebot/Plugins/Quote.hs (hunk 3 above)
Comment (L0neGamer) [reply src/Tablebot/Plugins/Quote.hs:382]: karl doesn't like short names and benji has word wrap on by default (I'll fix this)

File: src/Tablebot/Handler.hs (hunk 5 above)
Comment (L0neGamer) [reply src/Tablebot/Handler.hs:127]: if the environment variable is not present, application commands will not be generated. if the variable is present but is `"global"`, global commands will be created. else, the specific named server will be used for the application commands.

File: src/Tablebot/Utility/SmartParser.hs [hunk 6]
```diff
@@ -43,26 +83,65 @@ import Text.Megaparsec
...
 -- those to run the provided function with the arguments parsed and the message
 -- itself.
-class PComm commandty s where
-  parseComm :: commandty -> Parser (Message -> EnvDatabaseDiscord s ())
+class PComm commandty s context returns where
+  parseComm :: (Context context) => commandty -> Parser (context -> EnvDatabaseDiscord s returns)
 
--- As a base case, remove the spacing and check for eof.
-instance {-# OVERLAPPING #-} PComm (Message -> EnvDatabaseDiscord s ()) s where
+-- TODO: verify that all the parsers for PComm actually work
```
[the remark this answers is not in the document]
Comment (L0neGamer) [reply src/Tablebot/Utility/SmartParser.hs:89]: the status is "I think it all works but this stuff put me in a fugue state when I made it so I would love for someone to check properly"

File: src/Tablebot/Plugins/Quote.hs (hunk 4 above)
Comment (L0neGamer) [reply src/Tablebot/Plugins/Quote.hs:381]: so the issue is that our current subcommanding turns things into the normal text-parsing straight away, which means I've had to manually implement subcommands here.  Additionally, due to how interactions work, I've had to make the quote adding and quote editing "self referential", meaning that the "view quote" link is the bot's message itself, which is obviously a bit tricky.

I also have some rudimentary slash command auto complete implemented for `/quote show`, which isn't very clean either.

**Comment by L0neGamer** [comment]
finnbar you were right
I've started the tutorial and I've seen another way of doing this
I don't think I want to implement it right now cause it's even more internals work that I don't have the time to do right now, but the gist is if we create some mkCommand functions we could have the text parsers and the application command parser stuff processed at the same time, and put in the same data structure
which would hopefully reduce on some of the separation
and if we're really smart, we could do the subcommand stuff at the same time

File: tutorials/3.Interactions.md [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+
+```haskell
+-- current `echoHelp`
+echoHelp :: RestOfInput Text -> DatabaseDiscord MessageDetails
+echoHelp (ROI t) = return $ messageDetailsBasic t
+```
+
+We take away the restriction of `RestOfInput` for now, breaking `myecho`, but we'll fix it shortly. 
+
+We then change `myEcho` so that instead of having `(parseComm echoHelp)`, we instead have `(parseComm (echoHelp . unROI))`. This function is now the same as it originally was, and we can reuse our new `echoHelp` for our interactions.
```
Comment (finnbar) [inline tutorials/3.Interactions.md:125]: It would be nice to have a tiny mention of what `unROI` is and why it is necessary

File: tutorials/3.Interactions.md [hunk 8]
```diff
@@ -0,0 +1,348 @@
+# Making and managing Interactions
+
+A recent change in Discord added some pretty neat ways of interacting. Unfortunately, it requires a fair amount of fiddling, so let's go through how to use this!
```
Comment (finnbar) [inline tutorials/3.Interactions.md:3]: Maybe add a screenshot here showing what these interactions are, and what the user can expect to get out of this tutorial?

File: tutorials/3.Interactions.md [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    • In the expression:
+        makeApplicationCommandPair
+          "myecho" "echo your input" (echoHelp . unLabel)
+      In an equation for ‘myEchoInteraction’:
+          myEchoInteraction
+            = makeApplicationCommandPair
+                "myecho" "echo your input" (echoHelp . unLabel)
+```
+
+Oh dear that's hideous. Oh yeah, we forget to actually label anything!
```
Comment (finnbar) [inline tutorials/3.Interactions.md:158]: You've neglected to say what labelling _is_ at this point - another screenshot showing a slash command at work may help?

File: tutorials/3.Interactions.md [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+We can achieve this in one of two ways. Firstly, we could create a function which has the labels we want and then make the slash command like that; secondly, we could use type applications to add the labels we want here. I'm going to go with the latter in this case, but I'll show both here.
+
+```haskell
+myEchoInteraction1 :: Maybe ApplicationCommandRecv
+myEchoInteraction1 = makeApplicationCommandPair "myecho" "echo your input" echoHelp'
+    where
+        echoHelp' :: Labelled "message" "the message to echo back" Text -> DatabaseDiscord MessageDetails
```
Comment (finnbar) [inline tutorials/3.Interactions.md:166]: What is `Labelled`? Is it provided by Tablebot or discord-haskell? I'm guessing it's a constructor for a labelled argument, but it'd still be nice to see that in a sentence - it groups a value with some label.

File: tutorials/3.Interactions.md [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+
+Discord also offers user and message application commands, which we haven't created nice interfaces for just yet, but are usable if you do create them.
+
+## Components
+
+Another cool thing added was a variety of widgets and gizmos that bots can add to messages called "components". These also use the interaction system to process, but we've abstracted that again.
+
+First, let's decide what we want to do with this. How about we add to the basic `myping`, and make it so that there's a button that says "Ping!", which people can click and it'll reply "pong"?
+
+Yes it's contrived, I'm sorry.
```
Comment (finnbar) [inline tutorials/3.Interactions.md:203]: ~~man and I thought the _first_ tutorial was contrived lol~~

File: tutorials/3.Interactions.md [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+The action we use in response to the button will be `myPingAction` as it was before, so we feed that and `False` to `processComponentInteraction`. The `False` means that we send a message, instead of updating a message (see [More complex components](#more-complex-components)).
+
+The reason I chose "myping pingbutton" as the identifier before is because of how we process and distribute components. The first word of the unique identifier has to be the plugin name, and the second has to be identifier of the particular component being processed.
+
+```haskell
+myPingButton :: ComponentRecv
+myPingButton = ComponentRecv "pingbutton" (processComponentInteraction myPingAction False)
+```
+
+Now we just load `myPingButton` into `onComponentRecvs` in our plugin creation, run the bot again, run the command, and press our sparkly button, and we get a message from the bot saying "pong"!
```
Comment (finnbar) [inline tutorials/3.Interactions.md:230]: Please provide an example of this - I'm not exactly sure what you're referring to here.

**Review by finnbar — CHANGES_REQUESTED** [review]
I'm happy with the code and how it runs, I just have a few points for your tutorial. This can be merged once these are fixed, merge conflicts are resolved, and I quickly test the results of that.
